File: milo/frontend/static/common/js/builder.js
// Copyright 2023 The LUCI Authors.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

// Code for managing the opt-in banner for the new builder page.
// TODO(weiweilin): remove this once we turned down the old builder page.

$(document).ready(function() {
  'use strict';

  function setCookie(name, value) {
    const farFuture = new Date(2100, 0).toUTCString();
    document.cookie = `${name}=${value}; expires=${farFuture}; path=/`;
  }

  const newBuilderPageLink = $('#new-builder-page-link');
  newBuilderPageLink.on('click', (e) => {
    if (!e.metaKey && !e.shiftKey && !e.ctrlKey && !e.altKey) {
      setCookie('showNewBuilderPage', true);
    }

    return true;
  });

  if (!document.cookie.includes('askForBuilderPageFeedback=false')) {
    $('#feedback-bar').toggleClass('show', true);
  }

  $('#dismiss-feedback-bar').click(e => {
    $('#feedback-bar').toggleClass('show', false);

    const expires = new Date(Date.now() + 7 * 24 * 60 * 60 * 1000).toUTCString();
    document.cookie = `askForBuilderPageFeedback=false; expires=${expires}; path=/`;
    e.preventDefault();
  });

  $('#feedback-link').click(e => {
    const feedbackComment = encodeURIComponent(
`From Link: ${document.location.href}
Please enter a description of the problem, with repro steps if applicable.
`);
    const url = `https://bugs.chromium.org/p/chromium/issues/entry?template=Build%20Infrastructure&components=Infra%3ELUCI%3EUserInterface&labels=Pri-2,Type-Bug&comment=${feedbackComment}`;
    window.open(url);
    e.preventDefault();
  });

  // Keep the new builder page version up-to-date even when the user is using
  // the legacy builder page.
  navigator.serviceWorker?.register('/ui/ui_sw.js');
});

File: milo/internal/buildsource/builder.go
// Copyright 2017 The LUCI Authors.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package buildsource

import (
	"strings"

	"go.chromium.org/luci/common/errors"
	"go.chromium.org/luci/grpc/grpcutil"

	"go.chromium.org/luci/milo/internal/model"
)

// BuilderID is the universal ID of a builder, and has the form:
//
//	buildbucket/bucket/builder
type BuilderID string

// Split breaks the BuilderID into pieces.
//   - backend is always 'buildbucket'
//   - backendGroup is either the bucket or builder group
//   - builderName is the builder name.
//
// Returns an error if the BuilderID is malformed (wrong # slashes) or if any of
// the pieces are empty.
func (b BuilderID) Split() (backend, backendGroup, builderName string, err error) {
	toks := strings.SplitN(string(b), "/", 3)
	if len(toks) != 3 {
		err =
			grpcutil.InvalidArgumentTag.Apply(errors.Fmt("bad BuilderID: not enough tokens: %q", b))

		return
	}
	backend, backendGroup, builderName = toks[0], toks[1], toks[2]
	switch {
	case backend != "buildbucket":
		err =
			grpcutil.InvalidArgumentTag.Apply(errors.Fmt("bad BuilderID: unknown backend %q", backend))

	case backendGroup == "":
		err = grpcutil.InvalidArgumentTag.Apply(errors.New("bad BuilderID: empty backendGroup"))
	case builderName == "":
		err = grpcutil.InvalidArgumentTag.Apply(errors.New("bad BuilderID: empty builderName"))
	}
	return
}

// SelfLink returns LUCI URL of the builder.
func (b BuilderID) SelfLink(project string) string {
	return model.BuilderIDLink(string(b), project)
}
